### DISCUSSION TITLE
Searching Datatype in all opened Data Type Archives from an Analysis

### DISCUSSION THREAD

**Discussion by fmagin** [NationalSecurityAgency/ghidra#4247] [Q&A]
I am writing an analysis and during that I need to check if a datatype with a certain name already exists in any of the open data type archives. Specifically there is a high chance that the datatype is already defined in a file archive that is opened by default, like the `mac_osx` archive.

The `FindDataTypeScript` searches all open datatype managers via `tool.getService(DataTypeManagerService.class)` https://github.com/NationalSecurityAgency/ghidra/blob/fafbaacd5f7b057feb3684f06e26b7944899496c/Ghidra/Features/Base/ghidra_scripts/FindDataTypeScript.java#L72-L80

but the `tool` object isn't obviously available inside `ghidra.app.services.Analyzer#added(Program program, AddressSetView set, TaskMonitor monitor, MessageLog log)` and the supplied `Program.getDataTypeManager` will be only the `DataTypeManager` for the program itself.

@astrelsky has a snippet that searches the consumers of the `Program` object for a `PluginTool`, but this seems like a giant hack
```java
	public static PluginTool getTool(DomainObject obj) {
		for (Object o : obj.getConsumerList()) {
			if (o instanceof PluginTool) {
				return (PluginTool) o;
			}
		}
		return null;
	}
```
https://github.com/astrelsky/Ghidra-Cpp-Class-Analyzer/blob/933cc702fe7abcca62a5090dbc67b25f15e3c020/src/main/java/cppclassanalyzer/utils/CppClassAnalyzerUtils.java#L42-L49

**Comment by astrelsky** [comment]
It is a giant hack but it is guaranteed to be there. Since you are writing an `Analyzer` you could do it this way.

```java
private DataTypeManagerService getDataTypeManagerService(Program program) {
    AutoAnalysisManager man = AutoAnalysisManager.getAnalysisManager(program);
    if (man == null) {
        // unlikely edge case
        return null;
    }
    return man.getDataTypeManagerService();
}
```

**Comment by dev747368** [comment]
![6gd99n](https://user-images.githubusercontent.com/48332326/168654071-7c071f80-ef27-47dc-8ee3-0e2113614951.jpg)

**Reply by astrelsky** [reply]
Unlikely edge case.

This would imply that you are running a headless script and would have to manually open any additional datatype archives wouldn't it?

**Reply by dev747368** [reply]
Wouldn't have to be a headless script since just a plain import / analyze via headless could start off any analyzer that is present.

My meme was trying to hint at why its so hard to get ahold of the DTM service in an analyzer.  For a non-rigorous analyzer, sure, you can get by with the hackery, but a 100% solution will have to figure out some way of signalling (from the user to the analyzer) what archives to search that isn't GUI based.

**Reply by astrelsky** [reply]
> Wouldn't have to be a headless script since just a plain import / analyze via headless could start off any analyzer that is present.
> 
> My meme was trying to hint at why its so hard to get ahold of the DTM service in an analyzer. For a non-rigorous analyzer, sure, you can get by with the hackery, but a 100% solution will have to figure out some way of signalling (from the user to the analyzer) what archives to search that isn't GUI based.

I completely missed that. I took it more towards not having a tool available in headless mode instead of why it is difficult to get from an Analyzer.

**Reply by dev747368** [reply]
Its a pain for sure. For example, the pdb analyzer has to consult global app preferences for previously configured symbol search locations, and if you want to toggle the "allow remote" boolean, there is an example headless prescript (PdbSymbolServerExamplePrescript) you can use in a headless env.

The older pdb analyzer used to have calls to SystemUtilities.isInHeadlessMode() and popup GUI dialogs directly.

**Reply by ryanmkurtz** [reply]
> My meme was trying to hint...

Memes you have to explain are best memes!

**Reply by fmagin** [reply]
To give @dev747368 some credit here, the meme got the overall point across fairly well. But I also ran into a problem with @astrelsky [C++ plugin straight up crashing when you run the headless analyzer without activating the plugin via the GUI first](https://github.com/astrelsky/Ghidra-Cpp-Class-Analyzer/pull/41) so I was already familiar with his disregard for the headless analyzer :P

I am actually quite concerned about this working with the Headless Analyzer. The archive I am specifically interested in is https://github.com/PoomSmart/IDAObjcTypes, so I don't really need to search _every_ available archive. If it isn't in that specific one or the ProgramDataTypeManager then it is probably not available.
Is there some default location a GDT archive can be placed so I can easily open it from an Analyzer (that might be in headless mode)? I guess I can just ship the archive with my extension that provides the analyzer

**Reply by dev747368** [reply]
Some untested ideas:

1) Have a companion plugin (FooAnalyzerConfigurationPlugin), that adds an action to the popup menu of the DataTypeManager GUI that gives the user a chance to add or remove the selected DT archive from a list you write as a ghidra.framework.preferences.Preference property.  (con: if you record the path/name of the archive, the user could move it and break the association)
2) Have a folder in your ghidra project (ie. yourproject:/fooAnalyzerDataTypes) where the analyzer will fetch / open any archive placed there by the user.
3) Use a plain analyzer option (Analyzer.registerOptions) to let the user type the name of the archive you want.  Include a FooAnalyzerHeadlessPrescript.java that tweaks the analyzer options, and that the user can modify to their needs.  See FunctionIDHeadlessPrescript, and the getCurrentAnalysisOptionsAndValues().



> so I was already familiar with his disregard for the headless analyzer :P

😄

**Reply by astrelsky** [reply]
I prefer to keep my head thank you very much. Not sure doing development headless is a great idea. 

But in all seriousness it's mostly because I don't use it and wouldn't know that I broke it. If you hadn't put the link to that pr in the comment I would have been pulling my hair out looking for the bug because I had completely forgotten about it. I'm personally more interested in tools for when you have to roll up your sleeves and do things manually as opposed to automation.
